docs/data/data-grid/server-side-data/index.md

```diff
@@ -0,0 +1,269 @@
+  start: number | string; // first row index to fetch or cursor information
+  end: number; // last row index to fetch
+  groupFields: GridColDef['field'][]; // list of grouped columns (`rowGrouping`)
+}
+```
+
+And here's the `GetRowsResponse` object for reference:
+
+```tsx
+interface GetRowsResponse {
+  rows: GridRowModel[];
+  rowCount?: number; // optional: to reflect updates in total `rowCount`
+  pageInfo?: {
+    hasNextPage?: boolean; // optional: when row count is unknown/inaccurate, if `truncated` is set or rowCount is not known, data will keep loading until `hasNextPage` is `false`
+    truncated?: number; // optional: to reflect rowCount is inaccurate (will trigger `x-y of many` in pagination)
+  };
+}
+```
+
+#### Updating data
+
+If provided, the method `dataSource.updateRow` will be called with the `GridRowModel` object whenever the user edits a row. This method is optional and you can skip it if you don't need to update the data on the server. It will work in a similar way as the `processRowUpdate` prop.
+
+#### Data grid props
+
+These data grid props will work with the server-side data source:
+
+- `dataSource: DataSource`: the data source object that you need to implement
+- `queryClient: QueryClient`: optional: if not provided, a new instance of `QueryClient` will be created internally
+- `rows`: will be ignored, could be skipped when `dataSource` is provided
+- `rowCount`: will be used to identify the total number of rows in the grid, if not provided, the grid will use the _GetRowsResponse.rowCount_ value
+
+Props related to grouped data (`treeData` and `rowGrouping`):
+
+- `getGroupKey(row: GridRowModel): string`
+
+  will be used by the grid to group rows by their parent group
+  This effectively replaces `getTreeDataPath`.
+  Consider this structure:
+
+  ```
+  - (1) Sarah // groupKey 'Sarah'
+    - (2) Thomas // groupKey 'Thomas'
+  ```
+
+  When (2) is expanded, the `getRows` function will be called with group keys `['Sarah', 'Thomas']`.
+
+- `hasChildren?(row: GridRowModel): boolean`
+
+  Will be used by the grid to determine if a row has children on server
+
+- `getChildrenCount?: (row: GridRowModel) => number`
+
+  Will be used by the grid to determine the number of children of a row on server
+
+#### Existing server-side features
+
+The server-side `dataSource` will change a bit the way existing server-side features work. Whenever a valid data source is passed the features `filtering`, `sorting`, `pagination` will automatically be set to `server`.
+
+However, if for some reason, you want to explicitly work one or more of these features to work on `client`, you can do so by passing the `client` value to the feature. Here's an example:
+
+```tsx
+<DataGridPro
+  columns={columns}
+  dataSource={myCustomDataSource} // will set `filtering`, `sorting`, `pagination` to `server`
+  filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change
```

Comment:
I guess yes, it will work like how the client-side filter works today, i.e when the rows prop update, or `updateRows` is called, the filter is persisted and applied on the new rows, likewise if the data is refetched or new rows are fetched they'll be subjected to the active client-side filtering rules.

Although this won't be a need for many users, I'm sure there will be use cases where people will want this.

Reply:
I'm curious, what would be the use-case for this? I have trouble making sense of server data with client filtering.

Reply:
> Is this assuming that all the server rows are fetched?

No, actually it's adressing those rare use-cases when let's say filtering has yet not implemented on the server and the developer wants to keep it working on the client-side on the partially server loaded data. Tbh, I'm also not 100% sure about this. It has been asked as a [question](https://www.notion.so/Data-source-draft-db58b8b749154420a407f231c5528c1d?d=407dd9c21ac349e59fc0926afe0e4443&pvs=4#9c8f2af9613b48089c89fd0b12c4d17a) in the data source proposal, which made me think it might be a use-case requested by some users.

Reply:
I'm not sure if that use-case makes sense. Wouldn't it be confusing for end-users, that the fetching is server-side but the filtering is client-side? And couldn't it create weird UX events, e.g. if there is filtering + lazy-loading, and the lazy-loaded rows are filtered out as it loads?

I feel like it might be desirable to avoid/restrict this use-case.

Reply:
Yeah, good point around the possibly weird UX experience.
Alright, since we are anyways not sure about the need for this, let's not do it until a demand is there from some users for this specifically. 👍